pep-0705.rst

```diff
@@ -0,0 +1,259 @@
+
+
+.. highlight:: rst
+
+Abstract
+========
+
+:pep:`589` defines the structural type ``TypedDict`` for dictionaries with a fixed set of keys.
+As it defines a mutable type, it is difficult to correctly annotate methods which accept read-only parameters in a way that doesn't prevent valid inputs.
+This PEP proposes a type constructor ``typing.TypedMapping`` to support this use case.
+
+For example, given the following definition::
+
+    from typing import NotRequired, TypedMapping
+
+    class Movie(TypedMapping):
+        name: str
+        year: NotRequired[int | None]
+
+    def movie_string(movie: Movie) -> str:
+        if movie.get("year") is None:
+            return movie["name"]
+        else:
+            return f'{movie["name"]} ({movie["year"]})'
+
+A type checker should then consider this code type safe::
+
+    from typing import TypedDict
+
+    class MovieRecord(TypedDict):
+        name: str
+        year: int
+
+    movie: MovieRecord = {'name': 'Blade Runner',
+                          'year': 1982}
+    print(movie_string(movie))
+
+Note that currently the code above will fail to type check if we define ``Movie`` based on ``TypedDict``, and there is no replacement.
+
+Motivation
+==========
+
+Representing structured data using (potentially nested) dictionaries with string keys is a common pattern in Python programs. :pep:`589` allows these values to be type checked when the exact type is known up-front, but it is hard to write read-only code code that accepts more specific variants: for instance, where fields may be subtypes or restrict a union of possible types. This is an especially common issue when writing APIs for services, which may support a wide range of input structures, and typically do not need to modify their input.
+
+Take for example the function ``movie_string`` defined above. Suppose its type signature is defined using a ``TypedDict``::
+
+    from typing import NotRequired, TypedDict
+
+    class Movie(TypedDict):
+        name: str
+        year: NotRequired[int | None]
+
+    class MovieRecord(TypedDict):
+        name: str
+        year: int
+
+    def movie_string(movie: Movie) -> str:
+        if movie.get("year") is None:
+            return movie["name"]
+        else:
+            return f'{movie["name"]} ({movie["year"]})'
+
+Attempting to pass a ``MovieRecord`` into ``movie_string`` results (using mypy) in the error::
+
+    Argument 1 to "movie_string" has incompatible type "MovieRecord"; expected "Movie"
+
+This particular use case should be type-safe, but the type checker correctly stops the
+user from passing a ``MovieRecord`` into a ``Movie`` argument in the general case because
+the ``Movie`` class has mutator methods that could potentially allow the function to break
+the type constraints in ``MovieRecord`` by setting ``movie["year"] = None`` or ``del movie["year"]``.
```

Comment:
```suggestion
the type constraints in ``MovieRecord`` (e.g. by ``movie["year"] = None`` or ``del movie["year"]``).
```

Further break up this run-on sentence, and clearly delineate that these are examples.